compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1869,6 +1875,97 @@ private boolean isValidTypeForLength(int typeTag) {
         return false;
     }
 
+    private boolean isValidFreezeFunction(BLangInvocation iExpr, BLangBuiltInMethod function, BType type) {
+        if (iExpr.argExprs.size() > 0) {
+            dlog.error(iExpr.pos, DiagnosticCode.TOO_MANY_ARGS_FUNC_CALL, function.getName());
+            return false;
+        }
+
+        boolean dataType = types.isAnydata(type);
+
+        if (type.tag == TypeTags.NIL || (!dataType && !isNonAnyDataFreezeAllowedType(type))) {
+            return false;
+        }
+
+        if (dataType) {
+            iExpr.type = type;
+        } else {
+            if (type.tag == TypeTags.UNION) {
+                BUnionType unionType = (BUnionType) type;
+                if (unionType.getMemberTypes().stream().anyMatch(memType -> memType.tag == TypeTags.ERROR)) {
+                    iExpr.type = type;
+                } else {
+                    Set<BType> unionTypeMemTypes = new LinkedHashSet<>(unionType.memberTypes.size() + 1);
+                    unionTypeMemTypes.addAll(unionType.memberTypes);
+                    unionTypeMemTypes.add(symTable.errorType);
+                    iExpr.type = new BUnionType(null, unionTypeMemTypes, false);
+                }
+            } else {
+                iExpr.type = new BUnionType(null, new LinkedHashSet<BType>() {{ add(type); add(symTable.errorType); }},
+                                            false);
+            }
+        }
+//        iExpr.originalType = getSafeType(iExpr.type, iExpr); TODO: FIX
+        iExpr.originalType = iExpr.type;
+        resultType = types.checkType(iExpr, iExpr.type, expType);
+        return true;
+    }
+
+    private boolean isValidIsFrozenFunction(BLangInvocation iExpr, BLangBuiltInMethod function, BType type) {
+        if (iExpr.argExprs.size() > 0) {
+            dlog.error(iExpr.pos, DiagnosticCode.TOO_MANY_ARGS_FUNC_CALL, function.getName());
+            return false;
+        }
+
+        if (!isIsFrozenAllowedType(type)) {
+            return false;
+        }
+
+        iExpr.type = symTable.booleanType;
+        resultType = types.checkType(iExpr, iExpr.type, expType);
+        return true;
+    }
+
+    private boolean isNonAnyDataFreezeAllowedType(BType type) {
+        int typeTag = type.tag;
+        if (typeTag == TypeTags.ANY) {
+            return true;
+        }
+
+        if (types.isAnydata(type)) {
+            return true;
```

Comment:
This is for recursive checks. For example, the type of a map may not be anydata, but there is the possibility that it may hold only anydata values at a particular instance. Thus, we would recursively check the map's constraint type here.

e.g.,
```ballerina
map<(string, string|PersonObj)> m;
```